server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1623,58 +1579,31 @@ private void doGetRepositoryData(ActionListener<RepositoryData> listener) {
     }
 
     /**
-     * Puts the given {@link RepositoryData} into the cache if it is of a newer generation and only if the repository is not using
-     * {@link #bestEffortConsistency}. When using {@link #bestEffortConsistency} the repository is using listing to find the latest
-     * {@code index-N} blob and there are no hard guarantees that a given repository generation won't be reused since an external
-     * modification can lead to moving from a higher {@code N} to a lower {@code N} value which mean we can't safely assume that a given
-     * generation will always contain the same {@link RepositoryData}.
+     * Cache repository data if repository data caching is enabled.
      *
-     * @param serialized serialized RepositoryData to cache if newer than the cache contents or null if no data should be cached for the
-     *                   given generation
-     * @param generation repository generation of the given repository data
+     * @param repositoryData repository data to cache
+     * @param version        repository metadata version used when writing the data to the repository
      */
-    private void cacheRepositoryData(@Nullable BytesReference serialized, long generation) {
-        assert generation >= 0 : "No need to cache abstract generations but attempted to cache [" + generation + "]";
+    private void cacheRepositoryData(RepositoryData repositoryData, Version version) {
+        if (cacheRepositoryData == false) {
+            return;
+        }
+        final RepositoryData toCache;
+        if (version.onOrBefore(SnapshotsService.OLD_SNAPSHOT_FORMAT)) {
```

Comment:
Why do we drop shard gens if `≤ OLD_SNAPSHOT_FORMAT` rather than if `< SHARD_GEN_IN_REPO_DATA_VERSION`? Seems like this is a change in behaviour if the min node version is v7.5.1 or v7.5.2 (and all snapshots have known versions).

Also just thinking about the other stuff that would have been dropped but isn't any more? I guess the cluster/repo UUIDs are ok to cache but what about index gens? Can we at least assert that this is already `IndexMetaDataGenerations.EMPTY`?

Reply:
Fixed the version and added the assertion. There's nothing else in the repo data that we can't trust except for numeric generations in very strange corner cases (see other comment in a second).